src/platform_impl/linux/kms/event_loop.rs

```diff
@@ -0,0 +1,1461 @@
+        #[cfg(not(feature = "kms-ext"))]
+        let seat_name = "seat0";
+
+        // find_card_path uses `udev` to enumerate the cards that are currently available, and then
+        // choose the first (usually perferred) one
+        let card_path = std::env::var("WINIT_DRM_CARD")
+            .ok()
+            .map_or_else(|| find_card_path(seat_name), |p| Ok(Into::into(p)))?;
+
+        #[cfg(feature = "kms-ext")]
+        // Opening the card using our seat allows us to do so unprivallaged
+        let dev = seat
+            .open_device(&card_path)
+            .map_err(|e| {
+                OsError::new(
+                    line!(),
+                    file!(),
+                    platform_impl::OsError::KmsError(format!("failed to initialize DRM: {}", e)),
+                )
+            })?
+            .1;
+
+        #[cfg(not(feature = "kms-ext"))]
+        // Opening this card with no seat present means that we must have root
+        // (or be part of the `video` user group)
+        let dev = std::os::unix::prelude::IntoRawFd::into_raw_fd(
+            std::fs::OpenOptions::new()
+                .read(true)
+                .write(true)
+                .open(&card_path)
+                .map_err(|e| {
+                    OsError::new(
+                        line!(),
+                        file!(),
+                        platform_impl::OsError::KmsError(format!(
+                            "failed to initialize DRM: {}",
+                            e
+                        )),
+                    )
+                })?,
+        );
+        let drm = Card(std::sync::Arc::new(dev));
+
+        #[cfg(feature = "kms-ext")]
+        // Using our seat to open our input manager allows us to do so unprivallaged
+        let mut input = input::Libinput::new_with_udev(Interface(seat, HashMap::new()));
+        #[cfg(not(feature = "kms-ext"))]
+        // Opening our input manager with no seat means we must do so as root
+        // (or be part of the `input` user group)
+        let mut input = input::Libinput::new_with_udev(Interface);
+
+        input.udev_assign_seat(seat_name).unwrap();
+
+        // XKB allows us to keep track of the state of the keyboard and produce keyboard events
+        // very similarly to how a Wayland Compositor would.
+        let xkb_ctx = xkb::Context::new(xkb::CONTEXT_NO_FLAGS);
+        // Empty strings translates to "default" for XKB (in this context)
+        let keymap = xkb::Keymap::new_from_names(
+            &xkb_ctx,
+            "",
+            "",
+            "",
+            "",
+            std::env::var("WINIT_XKB_OPTIONS").ok(),
+            xkb::KEYMAP_COMPILE_NO_FLAGS,
+        )
+        .ok_or_else(|| {
+            OsError::new(
+                line!(),
+                file!(),
+                platform_impl::OsError::KmsMisc("failed to compile XKB keymap"),
+            )
+        })?;
+
+        let state = xkb::State::new(&keymap);
```

Comment:
I do remember that you've asked wrt what should be done with `xkb`. I was saying that you should split code from that file into separate modules, so you'd have input related stuff in input.rs, and so on, since right now this is quite a huge file to grasp.

The logical separation done which is done in wayland/x11 backends could be taken as an example.

Right now my main issue with this particular file that it contains all the logic for everything in winit, but if you split related code into separate files and be more declarative it would be much more helpful for others to read it.

Reply:
Fixed?